### PR TITLE
Do dtype conversion in torch hook to save memory

### PR DESCRIPTION
**Pull request by irenedea**
## Unit tests
Existing unit tests should cover this change. 

## Manual tests
Before
image

After

image

### CODE REVIEW

**Review by dakinggg — APPROVED** [review]
lgtm, can you get @bigning to take a look too please?

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 1]
```diff
@@ -436,7 +436,7 @@ def _save_checkpoint(self, state: State, logger: Logger):
         cpu_offload = True
 
         # Add a dtensor->cpu tensor hook to avoid CUDA OOM
```
Comment (snarayan21) [inline llmfoundry/callbacks/hf_checkpointer.py:438]:
```suggestion
        # Add hook to move tensors to cpu to avoid CUDA OOM
```

**Review by snarayan21 — APPROVED** [review]
lgtm! minor comment

File: llmfoundry/callbacks/hf_checkpointer.py [hunk 2]
```diff
@@ -449,20 +449,23 @@ def dtensor_to_tensor_hook(
...
+                del tensor
             if dist.get_global_rank() != 0:
-                for fqn in dtensor_fqns:
-                    del state_dict[fqn]
+                state_dict = {}
             return state_dict
 
         hooks = []
         for _, module in state_dict_model.named_modules():
-            if isinstance(module, FSDP):
```
Comment (bigning) [inline llmfoundry/callbacks/hf_checkpointer.py:462]: why do we need to remove this ? Does it make any difference ? without this condition, for modules wrapped with FSDP,  i think we run the `tensor_hook` for both the module and the fsdp wrapped module, which might be not necessary ?

**Review by bigning — COMMENTED** [review]
thanks for the change!

**Review by bigning — APPROVED** [review]

**Pull request merged**